### DISCUSSION TITLE
How do I configure deletion protection for devices? (PROTECTION_RULES)

### DISCUSSION THREAD

**Discussion by LHBL2003** [netbox-community/netbox#16021] [Q&A] [answered]
How do I configure deletion protection for devices?

Some time ago, I asked an FR https://github.com/netbox-community/netbox/issues/14996 to set up deletion protection for devices. I then learnt from jeremystretch that this should be possible with the following rule.

https://docs.netbox.dev/en/stable/configuration/data-validation/#protection_rules

As it happens, I didn't get round to setting it up and wanted to do it now. (Of course I deleted something by mistake again).

So I added this Protection Roles entry. If I now create a new device, I can delete it again immediately, although the status is "Active". At first I thought that my config was not working, but if I change something else in the config file and restart Netbox, then a change happens. 

So how do I configure PROTECTION_RULES so that devices are not deleted when the status is Active?
This is inserted here, isn't it?
$INSTALL_ROOT/netbox/netbox/configuration.py

Thanks for your info

**Comment by candlerb** [comment]
> So I added this Protection Roles entry

What exactly did you put there?

**Comment by LHBL2003** [comment]
This:

```
PROTECTION_RULES = {
    "dcim.site": [
        {
            "status": {
                "eq": "decommissioning"
            }
        },
        "my_plugin.validators.Validator1",
    ]
}
```

In the hope that you can only delete the devices if they are in the following status: decommissioning

**Reply by candlerb** [reply]
And what code did you provide for the custom validator, `my_plugin.validators.Validator1` ?

**Comment by LHBL2003** [comment]
I am a clueless fool when it comes to validation. :D

I was also puzzled by the line in the example, which is why I left it out. I just copied it one-to-one from this link and had no success with it. From the description, it sounds like that's what I need. Only the code that is given is a bit confusing.

https://docs.netbox.dev/en/stable/configuration/data-validation/#protection_rules

**Reply by candlerb** [reply]
What version of Netbox are you using? PROTECTION_RULES were only added in [v3.7.0](https://github.com/netbox-community/netbox/releases/v3.7.0)

Otherwise, perhaps your PROTECTION_RULES setting isn't being picked up, e.g. it went in the wrong file: it would normally be `/opt/netbox/netbox/netbox/configuration.py`. You can always configure it via the web UI instead under "Admin > Config Revisions"

> I was also puzzled by the line in the example, which is why I left it out

You can see a test case here: https://github.com/netbox-community/netbox/commit/edc4a35296b80d11851be49fbe3abaae8004e841#diff-a0fb2dfd5ca2af5b83efead658d613baea753bda6bfb7243e0b6f744e41b9592R195-R201

You don't need to provide any custom validator. But you *did* say you had included it in your config.

**Comment by LHBL2003** [comment]
I am currently using 3.7.1 snd yes, for me it is in the configuration file: /opt/netbox/netbox/netbox/configuration.py
At first I thought it was an installation error on my part because I have Netbox 3 times. Therefore I had given the link as in the example. (My mistake that has now led to confusion).

At least I knew that I had read validation somewhere on the website, I just couldn't remember where. I tried it once via the web frontend, but apparently I also entered it incorrectly there. 

Can you give me an example of how to spell it correctly? Because that doesn't work.

![image](https://github.com/netbox-community/netbox/assets/46369917/ad21b4ef-e6c3-4f43-b151-0411e7730301)

**Comment by LHBL2003** [comment]
I have now updated to 4.0.0. And in the Admin --> Configuration History before and after area I created a configuration with a maximum page size of 500. And I can still select 1000 in each list. I think the configurations are never used by me.

![image](https://github.com/netbox-community/netbox/assets/46369917/bd912f93-e8c3-44f5-8284-064f795f7b12)

![image](https://github.com/netbox-community/netbox/assets/46369917/3f04afda-80df-407e-8d48-b6e87e49dbbf)

![image](https://github.com/netbox-community/netbox/assets/46369917/93a6b192-338c-44ee-916d-7578ce229b1a)

**Comment by LHBL2003** [comment]
OK the config is partially loaded.
Then I just need to know how to enter the PROTECTION RULES correctly.

![image](https://github.com/netbox-community/netbox/assets/46369917/14230555-359d-4395-8867-c1083bcc4cba)

**Reply by candlerb** [reply, marked as answer]
From your original post:

> So I added this Protection Roles entry. If I now create a new device, I can delete it again immediately, although the status is "Active"

Your protection rule is configured to protect `dcim.site`, not `dcim.device`. Try adding and deleting a site instead.

**Comment by bitcollector1** [comment]
These protection rules are pretty great!  Thanks for the info folks!

**Comment by LHBL2003** [comment]
### Summary of the solution

"dcim.site, not dcim.device" oh man, I could have thought of that myself :D
I completely ignored that. Thanks @candlerb for your great help.

So here's a summary for everyone (Tested in version 4.0.0):

Admin --> Configuration Config Revisions (v3.x.x) --> Add (or Edit last Config)
Admin --> Configuration History (> v4.0.0) --> Add (or Edit last Config)

Protection rules
```
{
    "dcim.device": [
        {
            "status": {
                "eq": "decommissioning"
            }
        }
    ]
}
```

![image](https://github.com/netbox-community/netbox/assets/46369917/29e16d0d-2869-439c-a400-48a28625c256)

If you navigate to a device, rack or site, you will find the path in the top right-hand corner.
![image](https://github.com/netbox-community/netbox/assets/46369917/ab45fb9d-3482-49d2-b73b-f2254437e5ab)


Create a device and delete it:
Message: Deletion is prevented by a protection rule: ["Custom validation failed for status: ['Ensure this value is equal to decommissioning.']"]
![image](https://github.com/netbox-community/netbox/assets/46369917/77f0738f-9708-4a93-9a06-ae68c9b3c769)

I'll see if I can / may revise the article https://docs.netbox.dev/en/stable/configuration/data-validation/#protection_rules a little.
